rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -136,44 +165,271 @@ def consume_result(self, used_amount: FVal) -> None:
+            matched_acquisitions=matched_acquisitions,
+            is_complete=is_complete,
+        )
 
     def __len__(self) -> int:
-        return len(self._acquisitions)
+        return len(self._acquisitions_heap)
+
 
+class FIFOCostBasisMethod(BaseCostBasisMethod):
+    """
+    Accounting in FIFO (first-in-first-out) method.
+    https://www.investopedia.com/terms/f/fifo.asp
+    """
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = FVal(0)
 
-class FIFOAcquisitionsOrder(BaseAcquisitionsOrder):
-    """Accounting in FIFO (first-in-first-out) order"""
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.append(acquisition)
+        """Adds an acquisition to the `_acquisitions_heap` using a counter to achieve the FIFO order."""  # noqa: E501
+        heapq.heappush(self._acquisitions_heap, AssetAcquisitionHeap(self._count, acquisition))
+        self._count += 1
 
 
-class LIFOAcquisitionsOrder(BaseAcquisitionsOrder):
-    """Accounting in LIFO (last-in-first-out) order"""
+class LIFOCostBasisMethod(BaseCostBasisMethod):
+    """
+    Accounting in LIFO (last-in-first-out) method.
+    https://www.investopedia.com/terms/l/lifo.asp
+    """
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = FVal(0)
+
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.appendleft(acquisition)
+        """Adds an acquisition to the `_acquisitions_heap` using a negated counter to achieve the LIFO order."""  # noqa: E501
+        heapq.heappush(self._acquisitions_heap, AssetAcquisitionHeap(-self._count, acquisition))
+        self._count += 1
 
 
-class CostBasisEvents:
-    used_acquisitions: List[AssetAcquisitionEvent]
-    acquisitions_manager: BaseAcquisitionsOrder
-    spends: List[AssetSpendEvent]
+class HIFOCostBasisMethod(BaseCostBasisMethod):
+    """
+    Accounting in HIFO (highest-in-first-out) method.
+    https://www.investopedia.com/terms/h/hifo.asp
+    """
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        """
+        Adds an acquisition to the `_acquisitions_heap` using the negated amount
+        of the acquisition to achieve the HIFO order.
+        """
+        heapq.heappush(self._acquisitions_heap, AssetAcquisitionHeap(-acquisition.amount, acquisition))  # noqa: E501
 
-    def __init__(self, cost_basis_method: CostBasisMethod) -> None:
-        """This class contains data about acquisitions and spends. `acquisitions` field contains
-        custom Iterable that provides acquisitions in the order defined by `cost_basis_method`
+
+class AverageCostBasisMethod(BaseCostBasisMethod):
+    """
+    Accounting in Average Cost Basis(ACB) method.
+    https://www.investopedia.com/terms/a/averagecostbasismethod.asp
+    """
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = FVal(0)
```

Comment:
you forgot to use the constant here

```suggestion
        self._count = ZERO
```